File: app/src/main/java/com/pr0gramm/app/services/NavigationProvider.java
package com.pr0gramm.app.services;

import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;

import com.google.common.base.Optional;
import com.google.common.collect.FluentIterable;
import com.google.common.collect.ImmutableList;
import com.pr0gramm.app.ContextSingleton;
import com.pr0gramm.app.R;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.api.categories.ExtraCategoryApiProvider;
import com.pr0gramm.app.feed.FeedFilter;
import com.pr0gramm.app.feed.FeedType;
import com.pr0gramm.app.orm.Bookmark;

import org.immutables.value.Value;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindDrawable;
import butterknife.ButterKnife;
import rx.Observable;

import static com.google.common.base.Preconditions.checkState;
import static java.util.Collections.singletonList;
import static rx.Observable.combineLatest;
import static rx.Observable.just;

/**
 */
@ContextSingleton
public class NavigationProvider {
    private static final Logger logger = LoggerFactory.getLogger("NavigationProvider");

    private final Context context;
    private final UserService userService;
    private final InboxService inboxService;
    private final Settings settings;
    private final BookmarkService bookmarkService;
    private final SingleShotService singleShotService;

    @BindDrawable(R.drawable.ic_black_action_favorite)
    Drawable iconFavorites;

    @BindDrawable(R.drawable.ic_black_action_home)
    Drawable iconFeedTypePromoted;

    @BindDrawable(R.drawable.ic_black_action_stelz)
    Drawable iconFeedTypePremium;

    @BindDrawable(R.drawable.ic_black_action_trending)
    Drawable iconFeedTypeNew;

    @BindDrawable(R.drawable.ic_action_random)
    Drawable iconFeedTypeRandom;

    @BindDrawable(R.drawable.ic_category_controversial)
    Drawable iconFeedTypeControversial;

    @BindDrawable(R.drawable.ic_black_action_bookmark)
    Drawable iconBookmark;

    @BindDrawable(R.drawable.ic_action_email)
    Drawable iconInbox;

    @BindDrawable(R.drawable.ic_drawer_bestof)
    Drawable iconFeedTypeBestOf;

    @BindDrawable(R.drawable.ic_drawer_text_24)
    Drawable iconFeedTypeText;

    @BindDrawable(R.drawable.ic_black_action_upload)
    Drawable iconUpload;

    @BindDrawable(R.drawable.ic_drawer_kfav)
    Drawable iconKFav;

    private final Observable<Boolean> extraCategoryApiAvailable;

    @Inject
    public NavigationProvider(Activity activity, UserService userService, InboxService inboxService,
                              Settings settings, BookmarkService bookmarkService,
                              SingleShotService singleShotService,
                              ExtraCategoryApiProvider extraCategoryApi) {

        // we dont want to hold a reference to the activity
        this.context = activity.getApplicationContext();

        this.userService = userService;
        this.inboxService = inboxService;
        this.settings = settings;
        this.bookmarkService = bookmarkService;
        this.singleShotService = singleShotService;

        // inject the images
        ButterKnife.bind(this, activity);

        // estimate if we should show those extra categories
        this.extraCategoryApiAvailable = checkExtraCategoryApi(extraCategoryApi);
    }

    private Observable<Boolean> checkExtraCategoryApi(ExtraCategoryApiProvider extraCategoryApi) {
        return extraCategoryApi.get().ping().map(r -> true)
                .doOnError(err -> logger.error("Could not reach category api: {}", String.valueOf(err)))
                .onErrorResumeNext(just(false))
                .distinctUntilChanged()
                .doOnNext(allowed -> logger.info("Showing extra categories: {}", allowed))
                .onErrorResumeNext(just(false))
                .replay(1).autoConnect().share();
    }

    public Observable<List<NavigationItem>> navigationItems() {
        // observe and merge the menu items from different sources
        return combineLatest(ImmutableList.of(
                categoryNavigationItems(),

                userService.loginState()
                        .flatMap(ignored -> bookmarkService.get())
                        .map(this::bookmarksToNavItem),

                inboxService.unreadMessagesCount()
                        .map(this::inboxNavigationItem)
                        .map(ImmutableList::of),

                Observable.just(singletonList(getUploadNavigationItem()))
        ), args -> {
            ArrayList<NavigationItem> result = new ArrayList<>();
            for (Object arg : args)
                //noinspection unchecked
                result.addAll((List<NavigationItem>) arg);

            return result;
        });
    }

    /**
     * Adds the default "fixed" items to the menu
     */
    public List<NavigationItem> categoryNavigationItems(
            @Nullable String username, boolean extraCategory) {

        List<NavigationItem> items = new ArrayList<>();

        items.add(ImmutableNavigationItem.builder()
                .action(ActionType.FILTER)
                .filter(new FeedFilter().withFeedType(FeedType.PROMOTED))
                .title(getString(R.string.action_feed_type_promoted))
                .icon(iconFeedTypePromoted)
                .build());

        items.add(ImmutableNavigationItem.builder()
                .action(ActionType.FILTER)
                .filter(new FeedFilter().withFeedType(FeedType.NEW))
                .title(getString(R.string.action_feed_type_new))
                .icon(iconFeedTypeNew)
                .build());


        if (extraCategory) {
            if (settings.showCategoryControversial()) {
                items.add(ImmutableNavigationItem.builder()
                        .action(ActionType.FILTER)
                        .filter(new FeedFilter().withFeedType(FeedType.CONTROVERSIAL))
                        .title(getString(R.string.action_feed_type_controversial))
                        .icon(iconFeedTypeControversial)
                        .build());
            }

            if (settings.showCategoryRandom()) {
                items.add(ImmutableNavigationItem.builder()
                        .action(ActionType.FILTER)
                        .filter(new FeedFilter().withFeedType(FeedType.RANDOM))
                        .title(getString(R.string.action_feed_type_random))
                        .icon(iconFeedTypeRandom)
                        .build());
            }

            if (settings.showCategoryBestOf()) {
                items.add(ImmutableNavigationItem.builder()
                        .action(ActionType.FILTER)
                        .filter(new FeedFilter().withFeedType(FeedType.BESTOF))
                        .title(getString(R.string.action_feed_type_bestof))
                        .icon(iconFeedTypeBestOf)
                        .build());
            }

            if (settings.showCategoryText()) {
                items.add(ImmutableNavigationItem.builder()
                        .action(ActionType.FILTER)
                        .filter(new FeedFilter().withFeedType(FeedType.TEXT))
                        .title(getString(R.string.action_feed_type_text))
                        .icon(iconFeedTypeText)
                        .build());
            }
        }

        if (settings.showCategoryPremium()) {
            if (userService.isPremiumUser()) {
                items.add(ImmutableNavigationItem.builder()
                        .action(ActionType.FILTER)
                        .filter(new FeedFilter().withFeedType(FeedType.PREMIUM))
                        .title(getString(R.string.action_feed_type_premium))
                        .icon(iconFeedTypePremium)
                        .build());
            }
        }

        if (username != null) {
            items.add(ImmutableNavigationItem.builder()
                    .action(ActionType.FAVORITES)
                    .filter(new FeedFilter().withFeedType(FeedType.NEW).withLikes(username))
                    .title(getString(R.string.action_favorites))
                    .icon(iconFavorites)
                    .build());
        }

        return items;
    }

    /**
     * Returns the menu item that takes the user to the inbox.
     */
    private NavigationItem inboxNavigationItem(Integer unreadCount) {
        return ImmutableNavigationItem.builder()
                .title(getString(R.string.action_inbox))
                .icon(iconInbox)
                .action(ActionType.MESSAGES)
                .layout(R.layout.left_drawer_nav_item_inbox)
                .unreadCount(unreadCount)
                .build();
    }

    private List<NavigationItem> bookmarksToNavItem(List<Bookmark> entries) {
        if (singleShotService.firstTimeToday("bookmarksLoaded"))
            Track.bookmarks(entries.size());

        boolean premium = userService.isPremiumUser();
        return FluentIterable.from(entries)
                .filter(entry -> premium || entry.asFeedFilter().getFeedType() != FeedType.PREMIUM)
                .transform(entry -> {
                    Drawable icon = iconBookmark.getConstantState().newDrawable();
                    String title = entry.getTitle().toUpperCase();
                    return (NavigationItem) ImmutableNavigationItem.builder()
                            .title(title).icon(icon).bookmark(entry)
                            .action(ActionType.BOOKMARK)
                            .filter(entry.asFeedFilter())
                            .build();
                })
                .toList();
    }

    private Observable<List<NavigationItem>> categoryNavigationItems() {
        return combineLatest(
                userService.loginState().map(UserService.LoginState::name),
                extraCategoryApiAvailable.startWith(true),
                this::categoryNavigationItems);
    }

    /**
     * Returns the menu item that takes the user to the upload activity.
     */
    private NavigationItem getUploadNavigationItem() {
        return ImmutableNavigationItem.builder()
                .title(getString(R.string.action_upload))
                .icon(iconUpload)
                .action(ActionType.UPLOAD)
                .build();
    }

    /**
     * Short for context.getString(...)
     */
    private String getString(int id) {
        return context.getString(id);
    }


    @Value.Immutable
    public abstract static class NavigationItem {
        abstract public String title();

        /**
         * This is the basic icon for this navigation item. You might want to
         * tint it later on.
         */
        abstract public Drawable icon();

        /**
         * If this item is {@link com.pr0gramm.app.services.NavigationProvider.ActionType#FILTER} or
         * {@link com.pr0gramm.app.services.NavigationProvider.ActionType#BOOKMARK}, then it contains
         * a filter.
         */
        abstract public Optional<FeedFilter> filter();

        /**
         * If this is {@link com.pr0gramm.app.services.NavigationProvider.ActionType#BOOKMARK},
         * this method returns an optional containing the bookmark.
         */
        abstract public Optional<Bookmark> bookmark();

        abstract public ActionType action();

        public boolean hasFilter() {
            return filter().isPresent();
        }

        @Value.Default
        public int layout() {
            return R.layout.left_drawer_nav_item;
        }

        @Value.Default
        public int unreadCount() {
            return 0;
        }

        @Value.Check
        protected void check() {
            checkState(action() != ActionType.BOOKMARK || bookmark().isPresent(),
                    "A bookmark is required, if action() == BOOKMARK");

            checkState((action() != ActionType.FILTER && action() != ActionType.BOOKMARK) || filter().isPresent(),
                    "A filter is required, if action() in (FILTER, BOOKMARK)");
        }
    }

    public enum ActionType {
        FILTER, BOOKMARK, MESSAGES, UPLOAD, FAVORITES
    }
}

File: app/src/main/java/com/pr0gramm/app/services/NavigationProvider.kt
package com.pr0gramm.app.services

import android.app.Activity
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import androidx.annotation.DrawableRes
import androidx.appcompat.content.res.AppCompatResources
import com.pr0gramm.app.Logger
import com.pr0gramm.app.R
import com.pr0gramm.app.Settings
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.feed.FeedFilter
import com.pr0gramm.app.feed.FeedType
import com.pr0gramm.app.model.bookmark.Bookmark
import com.pr0gramm.app.model.config.Config
import com.pr0gramm.app.model.user.LoginState
import com.pr0gramm.app.orm.asFeedFilter
import com.pr0gramm.app.services.config.ConfigService
import com.squareup.picasso.Picasso
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.runInterruptible
import kotlin.time.Duration
import kotlin.time.ExperimentalTime

/**
 */
@OptIn(ExperimentalTime::class)
class NavigationProvider(
        private val context: Activity,
        private val userService: UserService,
        private val inboxService: InboxService,
        private val bookmarkService: BookmarkService,
        private val configService: ConfigService,
        private val singleShotService: SingleShotService,
        private val picasso: Picasso) {

    private val logger = Logger("NavigationProvider")

    private val iconBookmark by drawable(R.drawable.ic_action_bookmark)
    private val iconBookmarkTrending by drawable(R.drawable.ic_action_trending)
    private val iconCollections by drawable(R.drawable.ic_collection_yes)
    private val iconFeedTypeBestOf by drawable(R.drawable.ic_action_bestof)
    private val iconFeedTypeControversial by drawable(R.drawable.ic_action_controversial)
    private val iconFeedTypeNew by drawable(R.drawable.ic_action_new)
    private val iconFeedTypePremium by drawable(R.drawable.ic_action_follow_full)
    private val iconFeedTypePromoted by drawable(R.drawable.ic_action_promoted)
    private val iconFeedTypeRandom by drawable(R.drawable.ic_action_random)
    private val iconInbox by drawable(R.drawable.ic_action_email)
    private val iconUpload by drawable(R.drawable.ic_action_upload)

    private val iconSettings by drawable(R.drawable.ic_action_settings)
    private val iconContact by drawable(R.drawable.ic_action_feedback)
    private val iconInvite by drawable(R.drawable.ic_action_invite)
    private val iconFAQ by drawable(R.drawable.ic_action_faq)
    private val iconPremium by drawable(R.drawable.ic_action_premium)
    private val iconLogin by drawable(R.drawable.ic_action_login)
    private val iconLogout by drawable(R.drawable.ic_action_logout)

    // set value to true to trigger a refresh of the flow once.
    private val refreshAfterNavItemWasDeletedStateFlow = MutableStateFlow(false)

    private fun drawable(@DrawableRes id: Int): Lazy<Drawable> {
        return lazy {
            AppCompatResources.getDrawable(context, id)!!
        }
    }

    fun navigationItems(currentSelection: Flow<FeedFilter?>): Flow<List<NavigationItem>> {
        val loginStates = userService.loginStates

        val items = loginStates.flatMapLatest { loginState ->
            val rawSources = listOf(
                    remoteConfigItems(loginState, upper = true),

                    currentSelection.map { selection ->
                        categoryNavigationItems(selection, loginState.name)
                    },

                    remoteConfigItems(loginState, upper = false),

                    inboxService.unreadMessagesCount()
                            .map { listOf(inboxNavigationItem(it)) },

                    flowOf(listOf(uploadNavigationItem)),

                    bookmarkService.observe().combine(currentSelection) { bookmarks, selection ->
                        bookmarksToNavItem(selection, bookmarks)
                    },

                    flowOf(footerItems(loginState)))

            val sources = rawSources.map { source ->
                source.onStart { emit(listOf()) }.retryWhen { err, _ ->
                    logger.warn("Could not get category sub-items, retrying soon: ", err)
                    delay(Duration.seconds(5))
                    true
                }
            }

            // flatten generated lists into one list.
            combine(sources) { values -> values.flatMap { it } }
        }

        return refreshAfterNavItemWasDeletedStateFlow.flatMapLatest { items }.distinctUntilChanged()
    }

    private fun footerItems(loginState: LoginState): List<NavigationItem> {
        val items = mutableListOf<NavigationItem>()

        items += staticItemDivider
        items += staticItemSettings
        items += staticItemContact

        if (loginState.authorized) {
            items += staticItemInvites
        }

        items += staticItemFAQ

        if (!loginState.authorized || !loginState.premium) {
            items += staticItemPremium
        }

        if (loginState.authorized) {
            items += staticItemLogout
        } else {
            items += staticItemLogin
        }

        return items
    }

    /**
     * Adds the default "fixed" items to the menu
     */
    private fun categoryNavigationItems(selection: FeedFilter?, username: String?): List<NavigationItem> {
        fun makeItem(title: String, icon: Drawable, filter: FeedFilter, action: ActionType = ActionType.FILTER) = NavigationItem(
                action = action, title = title, icon = icon, filter = filter, isSelected = filter == selection)

        val items = mutableListOf<NavigationItem>()

        items += makeItem(
                title = getString(R.string.action_feed_type_promoted),
                icon = iconFeedTypePromoted,
                filter = FeedFilter().withFeedType(FeedType.PROMOTED))

        items += makeItem(
                title = getString(R.string.action_feed_type_new),
                icon = iconFeedTypeNew,
                filter = FeedFilter().withFeedType(FeedType.NEW))

        items += makeItem(
                title = getString(R.string.action_feed_type_bestof),
                icon = iconFeedTypeBestOf,
                filter = FeedFilter().withFeedType(FeedType.BESTOF))

        if (Settings.showCategoryControversial) {
            items += makeItem(
                    title = getString(R.string.action_feed_type_controversial),
                    icon = iconFeedTypeControversial,
                    filter = FeedFilter().withFeedType(FeedType.CONTROVERSIAL))
        }

        if (Settings.showCategoryRandom) {
            items += makeItem(
                    title = getString(R.string.action_feed_type_random),
                    icon = iconFeedTypeRandom,
                    filter = FeedFilter().withFeedType(FeedType.RANDOM))
        }

        if (Settings.showCategoryStalk) {
            items += makeItem(
                    title = getString(R.string.action_feed_type_premium),
                    icon = iconFeedTypePremium,
                    filter = FeedFilter().withFeedType(FeedType.STALK))
        }

        if (username != null) {
            items += makeItem(
                    action = ActionType.COLLECTIONS,
                    title = getString(R.string.action_collections),
                    icon = iconCollections,
                    filter = FeedFilter()
                            .withFeedType(FeedType.NEW)
                            .withCollection(username, "**ANY", "**ANY"))
        }

        return items
    }

    /**
     * Returns the menu item that takes the user to the inbox.
     */
    private fun inboxNavigationItem(unreadCounts: Api.InboxCounts): NavigationItem {
        return NavigationItem(
                action = ActionType.MESSAGES,
                title = getString(R.string.action_inbox),
                icon = iconInbox,
                layout = R.layout.left_drawer_nav_item_inbox,
                unreadCount = unreadCounts)
    }

    private fun bookmarksToNavItem(currentSelection: FeedFilter?, bookmarks: List<Bookmark>): List<NavigationItem> {
        val items = bookmarks
                .filter { userService.userIsPremium || it.asFeedFilter().feedType !== FeedType.STALK }
                .mapTo(mutableListOf()) { entry ->

                    val filter = entry.asFeedFilter()
                    val filterInverse = filter.invert()

                    val icon = when {
                        entry.trending -> iconBookmarkTrending.constantState!!.newDrawable()
                        else -> iconBookmark.constantState!!.newDrawable()
                    }

                    val layoutId = when {
                        entry.trending && filterInverse != null -> R.layout.left_drawer_nav_item_trending
                        else -> R.layout.left_drawer_nav_item
                    }

                    val title = entry.title

                    val isSelected = filter == currentSelection || filterInverse == currentSelection

                    NavigationItem(
                            action = ActionType.BOOKMARK, layout = layoutId,
                            title = title, icon = icon, bookmark = entry,
                            filter = filter, filterInverse = filterInverse,
                            isSelected = isSelected)
                }

        val actionKey = "hint:bookmark-hold-to-delete:2"

        val hintNotYetShown = singleShotService.isFirstTime(actionKey)
        if (items.isNotEmpty() && bookmarkService.canEdit && hintNotYetShown) {
            val hint = NavigationItem(
                    action = ActionType.HINT,
                    title = getString(R.string.bookmark_hint_delete),
                    icon = null,
                    layout = R.layout.left_drawer_nav_item_hint,
                    customAction = {
                        singleShotService.markAsDoneOnce(actionKey)
                        refreshAfterNavItemWasDeletedStateFlow.value = true
                    })

            items.add(0, hint)
        }

        if (items.isNotEmpty()) {
            items.add(0, staticItemDivider)
        }

        return items
    }

    /**
     * Returns the menu item that takes the user to the upload activity.
     */
    private val uploadNavigationItem: NavigationItem by lazy {
        NavigationItem(
                action = ActionType.UPLOAD,
                title = getString(R.string.action_upload),
                icon = iconUpload)
    }

    /**
     * Divider to divide item groups
     */
    private val staticItemDivider: NavigationItem by lazy {
        NavigationItem(ActionType.DIVIDER,
                layout = R.layout.left_drawer_nav_item_divider)
    }

    private fun staticItem(action: ActionType, icon: Drawable, title: String): NavigationItem {
        return NavigationItem(action, title = title, icon = icon, colorOverride = 0x80808080.toInt())
    }

    private val staticItemFAQ: NavigationItem by lazy {
        staticItem(ActionType.FAQ, iconFAQ, getString(R.string.action_faq))
    }

    private val staticItemSettings: NavigationItem by lazy {
        staticItem(ActionType.SETTINGS, iconSettings, getString(R.string.action_settings))
    }

    private val staticItemInvites: NavigationItem by lazy {
        staticItem(ActionType.INVITES, iconInvite, getString(R.string.action_invite))
    }

    private val staticItemContact: NavigationItem by lazy {
        staticItem(ActionType.CONTACT, iconContact, getString(R.string.action_contact))
    }

    private val staticItemPremium: NavigationItem by lazy {
        staticItem(ActionType.PREMIUM, iconPremium, getString(R.string.action_premium))
    }

    private val staticItemLogin: NavigationItem by lazy {
        staticItem(ActionType.LOGIN, iconLogin, getString(R.string.action_login))
    }

    private val staticItemLogout: NavigationItem by lazy {
        staticItem(ActionType.LOGOUT, iconLogout, getString(R.string.action_logout))
    }

    /**
     * Short for context.getString(...)
     */
    private fun getString(id: Int): String {
        return context.getString(id)
    }

    private fun remoteConfigItems(loginState: LoginState, upper: Boolean): Flow<List<NavigationItem>> {
        return configService.observeConfig()
                .map { config -> config.specialMenuItems }
                .distinctUntilChanged()
                .flatMapLatest { item -> resolveRemoteConfigItems(loginState, upper, item) }

    }

    /**
     * Resolves the special for the given config and loads the icon,
     * if there is one.
     */
    private fun resolveRemoteConfigItems(
            loginState: LoginState, upper: Boolean,
            items: List<Config.MenuItem>): Flow<List<NavigationItem>> {

        val images = flow {
            val itemsToLoad = items.filterNot { item ->
                item.requireLogin && !loginState.authorized || item.lower != !upper
            }

            val navItems = runInterruptible(Dispatchers.IO) {
                itemsToLoad.map { item ->
                    logger.debug { "Loading item $item" }

                    val bitmap = picasso.load(Uri.parse(item.icon))
                            .noPlaceholder()
                            .resize(iconUpload.intrinsicWidth, iconUpload.intrinsicHeight)
                            .get()

                    logger.info { "Loaded image for $item" }
                    val icon = BitmapDrawable(context.resources, bitmap)
                    val uri = Uri.parse(item.link)

                    val layout = if (item.noHighlight) R.layout.left_drawer_nav_item else R.layout.left_drawer_nav_item_special
                    NavigationItem(ActionType.URI, item.name, icon, uri = uri, layout = layout)
                }
            }

            emit(navItems)
        }

        return images.retry(retries = 3).catch { err ->
            logger.warn(err) { "Could not load item after 3 retries." }
        }
    }

    class NavigationItem(val action: ActionType,
                         val title: CharSequence? = null,
                         val icon: Drawable? = null,
                         val layout: Int = R.layout.left_drawer_nav_item,
                         val filter: FeedFilter? = null,
                         val filterInverse: FeedFilter? = null,
                         val bookmark: Bookmark? = null,
                         val unreadCount: Api.InboxCounts = Api.InboxCounts(),
                         val customAction: () -> Unit = {},
                         val uri: Uri? = null,
                         val isSelected: Boolean = false,
                         val colorOverride: Int? = null) {

        private val hashCode by lazy {
            listOf(action, title, layout, filter, bookmark, unreadCount, uri, isSelected).hashCode()
        }

        override fun hashCode(): Int = hashCode

        override fun equals(other: Any?): Boolean {
            return other is NavigationItem
                    && other.action == action
                    && other.title == title
                    && other.layout == layout
                    && other.filter == filter
                    && other.bookmark == bookmark
                    && other.unreadCount == unreadCount
                    && other.uri == uri
                    && other.isSelected == isSelected
        }
    }

    enum class ActionType {
        HINT, FILTER, BOOKMARK, MESSAGES, UPLOAD, COLLECTIONS, URI,
        DIVIDER,
        SETTINGS, CONTACT, INVITES, FAQ, PREMIUM, LOGIN, LOGOUT
    }
}
